src/ray/util/mutex_protected.h

```diff
@@ -0,0 +1,70 @@
+// Copyright 2024 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <type_traits>
+#include <utility>
+
+#include "absl/synchronization/mutex.h"
+
+namespace ray {
+
+// A wrapper class that protects a value with a mutex. One can get a const& with a reader
+// lock and a mutable & with a writer lock.
+//
+// Limitation: we only protect a single value. For the classes where a mutex protects
+// multiple values, or with methods that requires/excludes locks, we have to write an
+// impl class to be protected.
+template <typename T>
+class MutexProtected {
+ public:
+  explicit MutexProtected(T initialValue) : value_(std::move(initialValue)) {}
+
+  // Default constructor enabled only if T is default-constructible
+  template <typename = typename std::enable_if_t<std::is_default_constructible_v<T>>>
```

Comment:
I'm not sure how useful the SFINAE is here, since if T not default constructible, compilation will fail.
Eg: https://onlinegdb.com/snTdrSZRW

Reply:
My point:
- SFINAE is used to provide constraint to template arguments, *usually for override selection*; I don't think it's a necessity here;
- SFINAE is notorious for hard-to-understanding, it adds mind overhead;
- The real improvement comes from using clang over gcc; clang shows significantly better error message than gcc